### ISSUE TITLE
POS printing

### ISSUE THREAD

**Issue by DirkRoelandt** [Microsoft/Windows-universal-samples#117]
I have an EPSON TM-T88IV with Ethernet interface. Works fine with Epson epos-print SDK in Windows Store App (under 8.1 and 10).
On same computer with the c# Windows 10 sample (https://github.com/Microsoft/Windows-universal-samples/tree/master/Samples/PosPrinter) it throws me an exception during in this statement : 
DeviceInformationCollection deviceCollection = await DeviceInformation.FindAllAsync(PosPrinter.GetDeviceSelector());
So my question is : where can I get the correct setup, driver, etc.. for this new suggested working method ? Adding a printer via control panel only finds the IP interface of the printer, leaving me clueless about the driver to take (surely it is not the suggested Epson ESC/P R printer).

**Comment by nbickel** [comment]
Have you found out something new? I also try to print to a Epson PosPrinter from my Windows 10 App.

At https://msdn.microsoft.com/en-us/library/windows/apps/mt426652 in the "Remarks"-Section are some supported printers listed. But they are also not working.

When i start the PosPrinter example i can go past the FindAllAsync() and find in one case a Bluetooth Epson Mobillink P20 and in another case an over Ethernet connected Epson TM-T70II but  at PosPrinter.FromIdAsync(deviceInfo.Id) I always get an Exception. At the moment I assume that there are now additional driver needed. But it is almost impossible to get more information. The printers are discoverd at DeviceInformation.FindAllAsync correctly without any additional Epson driver. 

I can choose to pair the device and get a message that the connection was successfull after that the Exception occur.

For the Mobillink P20 I get a "Ein Verbindungsversuch ist fehlgeschlagen, da die Gegenstelle nach einer bestimmten Zeitspanne nicht richtig reagiert hat, oder die hergestellte Verbindung war fehlerhaft, da der verbundene Host nicht reagiert hat. (Exception from HRESULT: 0x8007274C)".

For the TM-T70II I get a "Zugriff verweigert (Exception from HRESULT: 0x80070005 (E_ACCESSDENIED))".

If someone at Microsoft needs more information to reproduce this issue let me know.

**Comment by DirkRoelandt** [comment]
Hi, 

No progess yet (downloaded the new samples of a few days ago).  Seems that your P20 and TM70 works "better" (although only a little bit).
My error was Exception thrown is Exception from HRESULT: 0x80004004 (E_ABORT)) in the indicated line.
Clearly there is some work to be done to let POS printers really let work under UWP.  
Help from Microsoft is indeed highly appreciated.

Did you install any driver (eg like with normal Windows printer on OS level ?).  If so, where did you find the Windows driver ?   Adding a printer only works until the discovery of the IP interface of POS printer.

**Comment by TerryWarwick** [comment]
The one piece of information that may not be obvious today is that you need to first the device through Start - Settings - Devices - Connected devices.  This is the case for both network and Bluetooth devices at this time.  Once you have paired the device then the sample should work. 

Please let us know if this addresses the problem you are encountering or not.

Terry Warwick
Microsoft

**Comment by DirkRoelandt** [comment]
Suggested path discovers an unknown device with as description (00-00-48-6b-56-10) - i.e. the correct Mac address of POS printer.  

But the sample still crashes in exact the same line.

The value returned by PosPrinter.GetDeviceSelector() is :

( System.Devices.DevObjectType:=10 AND System.Devices.AepService.ServiceClassId:="{C7BC9B22-21F0-4F0D-9BB6-66C229B8CD33}" AND System.Devices.AepService.ProtocolId:="{D4BF61B3-442E-4ADA-882D-FA7B70C832D9}" AND System.Devices.AepService.ParentAepIsPaired:=System.StructuredQueryType.Boolean#False ) OR ( System.Devices.DevObjectType:=8 AND System.Devices.InterfaceClassGuid:="{C7BC9B22-21F0-4F0D-9BB6-66C229B8CD33}" AND System.Devices.InterfaceEnabled:=System.StructuredQueryType.Boolean#True )

**Comment by nbickel** [comment]
I can confirm that i can connect a TM-T70II over WLAN with "Start - Settings - Devices - Connected devices" and print the demo receipt. 

The same is not possible for the Mobillink P20 over Bluetooth. The Bluetooth device gets added to the Bluetooth connected devices but there is still a "connection takes too long"-Exception at "PosPrinter.FromIdAsync(deviceInfo.Id)" in the demo.

**Comment by hannespreishuber** [comment]
as the supported devices are quite expensive and there are USB printer for less 20$ my question is, why is there not a universal text driver-

**Comment by TerryWarwick** [comment]
nbickel,

Good to hear you have been able to connect and print to the TM-T70II over WLAN.  Please make sure you are initiating the pairing of your Bluetooth Mobillink P20 through Connected Devices verses Bluetooth as well.

Terry Warwick
Microsoft

**Comment by TerryWarwick** [comment]
Hannespreishuber,

Our POSPrinter implementation for Windows 10 is focused on enabling tablets and phone form factors to print to typical thermal receipt printers connected via network or Bluetooth.  You can also use this implementation for stationary point of sale systems provided that POSPrinter is connected over the network or Bluetooth.  We chose ESC/POS printer control language because the vast majority of POSPrinters on the market support this PCL either by default or through emulation.

You can still print to USB connected printers through either a Windows driver that supports the specific printer, or through one of the UnifiedPOS implementations (OPOS, JavaPOS, POS for .NET) from a legacy Win32 application.

If you are using another USB printer that supports text only output, you can add the Windows 10 Generic Text Only printer driver.  
1. Go to Control Panel - Devices and Printers then select Add a Printer.  
2. Click on "The printer that I want isn't listed"
3. Select "Add a local printer or network printer with manual settings" then select Next
4. Enter the port your printer is connected to then select next
5.  Choose Generic from the Manufacturer list 
6. Select Generic / Text Only from the Printers list then select Next
7.  Give the printer a name and select Next

Terry Warwick
Microsoft

**Comment by DirkRoelandt** [comment]
Terry,

Following today's feedback, my input : 

1.
FYI : the TM88 (is on the list for supported printer - https://msdn.microsoft.com/en-us/library/windows/apps/mt426652.aspx) that doesn't work is connected via LAN.

2. 
Do hope that the support will be extended to ESC/P printers with eg USB interface.  Target market for Windows 10 UWP only supporting IP printing is too limited (a lot - majority actually - of POS terminals are and will still be on "desktop" hardware).

KR.

Dirk

**Comment by hannespreishuber** [comment]
Terry -thanks that is clear- I do it with winforms or wpf like that.
But the pos printer api looks much simpler - and I want to dev for future and multiple devices...

**Comment by TerryWarwick** [comment]
You should be able to print to ESC/P printers using standard printer drivers from Epson.  This API is strictly for printing to standard receipt printers. You might also want to check out the Printing sample for more guidance on how to print to non-POSPrinter printers.  

https://github.com/Microsoft/Windows-universal-samples/tree/master/Samples/Printing

For Win32 "desktop" apps you can still use the UnifiedPOS implementations (mentioned earlier), which are similar to our UPW solution.  We agree that USB is important, but for Win10 RTM release we felt that it was even more important to focus on addressing the mobile space.  We do appreciate your feedback and will take this into account in our planning for future updates.  

We have not been able to reproduce the problems that you are encountering with the Epson TM-T88IV and we are investigating.  Is the receipt printer connected to a hub that your PC is also connected to? I would like to understand the specific physical connections between the PC and the printer to see if it will help us understand why you are unable to talk to the printer using our sample app.  Also please make sure that you have reset your connections from when you were using the Epson SDK to talk to the device.  We need to make sure the two different solutions are not competing for a lock on the printer.

Terry Warwick
Microsoft

**Comment by DirkRoelandt** [comment]
The printer is connected directly to PC.  PC ethernet set to fix IP address in range for printer's. All other programs are not running.

What I already find strange is that only the MAC appear after the pairing.

These are the exact details from the printer : firmware version 10.03 ESC/POS, Hardware version 02.00, Software version 02.22

Setting up printer via the Generic / Text Only  driver : works for the test print and any print for eg word.
Although Printer sample does not work (even not eg Microsoft XPS Document View, Microsoft Print to PDF).

**Comment by TerryWarwick** [comment]
Dirk,

We have confirmed that we can print to an Epson TM-T88IV that is directly connected to an Ethernet port on the PC by hardcoding the IP address into the range for the printer.

At a command prompt, can you please run the command ver and send me the details.  With so many people in the Windows Insider program, I just want to confirm the version of Windows that you are running.

We would like to collect an event log from you and I am investigating a way for you to transfer the file to me, but in the mean time can you follow these steps to create the logfile we need?

1.  Create a file named "providers.txt" with the following content
         {3C27A357-A925-4040-87E4-82E74EF8731E} 255 255
         {a1e769b6-694b-46d3-a804-3a5251d9527e} 255 255

2. In an admin command prompt, run the following command:
         Logman start postrace –pf providers.txt –o postrace.etl -ets

3. Now reproduce the failure that you are encountering

4. Again in an admin command prompt, run the following command:
         Logman stop postrace –ets

This will have created the file postrace.etl

I'll get back to you shortly on sending the file to me, I am confirming the process with our support staff.

Terry Warwick
Microsoft

**Comment by nbickel** [comment]
When i connect my Mobilink-P20 over Connected Devices it gets added to the Bluetooth devices but not to the List of "Connected Devices". After that i can find the printer via FindAllAsnyc but i can't get it with FromIdAsync. Anythink more i can try?

In Europe the Mobilink-P20 is sold under the name TM-P20 but Epson support assured me that it is just a different name. Since i can find the printer with FindAllAsnyc i think it is not a problem of "detection". More of a connection issue.

**Comment by TerryWarwick** [comment]
Dirk,

Can you update your providers.txt with the following and send me a new log using the steps provided earlier?

-------
{3C27A357-A925-4040-87E4-82E74EF8731E} 0xFFFFFFFFFFFFFFFF 255
{a1e769b6-694b-46d3-a804-3a5251d9527e} 0xFFFFFFFFFFFFFFFF 255
{19e464a4-7408-49bd-b960-53446ae47820} 0xFFFFFFFFFFFFFFFF 255
-------

Terry Warwick
Microsoft

**Comment by TerryWarwick** [comment]
nbickel,

We are currently investigating a connection issue with the P20. Some, but not all P20's are having the same issue that you are encountering.  We are working with Epson to identify and address this issue.  I apologize for the inconvenience.

Terry Warwick
Microsoft

**Comment by TerryWarwick** [comment]
Dirk,
Unfortunately, the log file is not helping us narrow down the problem.  Is there anything else about your configuration that we should be aware of?  Did this PC ever have any Windows Insider builds on it prior to the final public build?  

We have tried to repro this on 3 different PC's each with different network connections using an Epson TM-T88V, but still cannot reproduce the problem that you are encountering.

Can you provide the DIP switch settings that your TM-T88IV currently has configured along with the revisions associated with the printer?  This should be on a small sticker affixed to the bottom of the printer.

Terry Warwick
Microsoft

**Comment by TerryWarwick** [comment]
Dirk,
We have been unable to acquire a TM-T88IV to troubleshoot this issue either directly from Epson or our device suppliers and have recently learned that the ESC/POS implementation on the TM-T88IV is somewhat different from the TM-T88V.  Unfortunately this forces us to remove the TM-T88IV from our compatibility list.  We would certainly revisit this decision in the future if we are able to obtain a TM-T88IV for further investigation and support validation.  I apologize for any confusion that this has created.

Terry Warwick
Microsoft

**Comment by GregorSpagnolo** [comment]
I connected the printer firstly via Bluetooth and I was not able to connect to it via store API. Than I connected it via USB cable. In both cases it was listed under Unspecified section.
![capture](https://cloud.githubusercontent.com/assets/4869477/11158427/751889e2-8a58-11e5-90af-bf6c153bd0bd.PNG)

I tried to call 
DeviceInformationCollection deviceCollection = await DeviceInformation.FindAllAsync();
the USB Portable Printer is inside the deviceCollection but when I call
PosPrinter posPrinter = await PosPrinter.FromIdAsync(deviceInfo.Id);
the posPrinter object is null.

Non W store apps works as expected

**Comment by TerryWarwick** [comment]
Gregor,

This thread is closed.  For new issues, please submit a new thread and provide details about the issue you are encountering, including the device you are working with.  Please connect to the device through Connected Devices UX rather than Bluetooth UX.  We know this is not intuitive and will be looking at addressing this in the future.

**Comment by wizarrc** [comment]
@TerryWarwick

> If you are using another USB printer that supports text only output, you can add the Windows 10 Generic Text Only printer driver.

Following up on your suggestion, I tried what you said to run the sample installing a text mode driver that prints to a file port for testing purposes and I can't seem to get a non-null value from PosPrinter.FromIdAsync. Do you have an example of how to glue the generic text driver to the PosPrinter class? My novice attempt below only retrieves a null value. If this is non-typical I can open up a new issue, but I posted here to give you context.

Scenario1_ReceiptPrinter.xaml.cs (modified)
```
const string GenericPrinterGuid = "63CC356A-77AF-4628-B529-B95B2C6EF952";
private async Task<bool> FindReceiptPrinter()
{
    if (printer == null)
    {
        rootPage.NotifyUser("Finding printer", NotifyType.StatusMessage);
        DeviceInformationCollection deviceCollection = await DeviceInformation.FindAllAsync(/*PosPrinter.GetDeviceSelector()*/);
        if (deviceCollection != null && deviceCollection.Count > 0)
        {
            DeviceInformation deviceInfo = deviceCollection.SingleOrDefault(y => y.Id.Contains(GenericPrinterGuid)); //= deviceCollection[0];
            //deviceInfo.Id is "\\?\SWD#PRINTENUM#{63CC356A-77AF-4628-B529-B95B2C6EF952}#{0ecef634-6ef0-472a-8085-5ad023ecbccd}"
            printer = await PosPrinter.FromIdAsync(deviceInfo.Id); //This is where it fails returning null
            if (printer != null)
            {
                if (printer.Capabilities.Receipt.IsPrinterPresent)
                {
                    rootPage.NotifyUser("Got Printer with Device Id : " + printer.DeviceId, NotifyType.StatusMessage);
                    return true;
                }
            }
            else
            {
                rootPage.NotifyUser("No Printer found", NotifyType.ErrorMessage);
                return false;
            }
        }
        else
        {
            rootPage.NotifyUser("No devices returned by FindAllAsync.", NotifyType.ErrorMessage);
            return false;
        }
    }
    return true;
}
```

**Comment by TerryWarwick** [comment]
@wizarrc 

POSPrinter does not interact with the Generic Text Only printer driver, but rather only ESC/POS remote printers today per the documentation on MSDN here:  [Windows.Devices.PointOfService.POSPrinter](https://msdn.microsoft.com/en-us/library/windows/apps/mt426652.aspx)

You may be able to use Generic Text Only printer driver or a driver from the printer manufacturer, but this would be a different path than using Windows.Devices.PointOfService.POSPrinter.

Terry Warwick
Microsoft

**Comment by wizarrc** [comment]
@TerryWarwick I must have misunderstood your reply in https://github.com/Microsoft/Windows-universal-samples/issues/117#issuecomment-132713638 about a universal print driver in the context of the POSPrinter.

Thanks for the quick reply.

**Comment by wizarrc** [comment]
@TerryWarwick

Am I to assume that the list of supported printers listed on the MSDN site is from a Microsoft supplied driver that automatically installs? I contacted the manufacturer (Epson) and they are not aware of any such Universal drivers that work in the PointOfService namespace of the Windows Runtime APIs. I did some digging and I see a windows driver in the catalog that would get installed automatically (%windir%\System32\drivers\UMDF\RemotePosDrv.dll) that matches the device selector in https://github.com/Microsoft/Windows-universal-samples/issues/117#issuecomment-132490478

Can you confirm that this is the case?

**Comment by TerryWarwick** [comment]
@wizarrc 

We have an in-box driver that supports any remote (wired or wireless network and bluetooth) POSPrinter that uses the ESC/POS printer control language. Please refer to the following documentation for more information:

[Windows.Devices.PointOfService](https://msdn.microsoft.com/en-us/library/windows/apps/windows.devices.pointofservice.aspx)
Enables application developers to access Point of Service (POS) peripheral devices. The namespace provides a vendor-neutral interface for accessing POS devices from various manufacturers from the 

[Out of Band Pairing](https://msdn.microsoft.com/en-us/windows/uwp/devices-sensors/out-of-band-pairing)
Out-of-band pairing allows apps to connect to a Point-of-Service peripheral without requiring discovery.

[Epson ESC/POS with Formatting](https://msdn.microsoft.com/en-us/windows/uwp/devices-sensors/epson-esc-pos-with-formatting)
Learn how to use the ESC/POS command language to format text, such as bold and double size characters, for your Point of Service printer.

Terry Warwick
Microsoft

**Comment by antalistvan** [comment]
I don't understand it.

How can I use our EPSON TM-T20ii printer from UWP (desktop PC) in case of using USB for the printer connection?

Look like there is no support from MS or EPSON for this case.

Am I right?

**Comment by TerryWarwick** [comment]
@antalistvan 

As stated in our documentation for [windows.devices.pointofservice.posprinter](https://msdn.microsoft.com/en-us/library/windows/apps/mt426652.aspx) on MSDN, Windows 10 1507, 1511 and 1607 releases only support remote printers connected via ethernet, Wi-Fi or Bluetooth for printers that support the ESC/POS printer control language.

You could try to interact with your USB printer from a UWP application through legacy driver support provided by OPOS or POS for .NET via [Desktop Application Converter](https://msdn.microsoft.com/en-us/windows/uwp/porting/desktop-to-uwp-run-desktop-app-converter), also known as Project Centennial.  In this case, the legacy stack along with service objects would need to be captured in your application package.

Terry Warwick
Microsoft

**Comment by DirkRoelandt** [comment]
Hi Terry, 

I have written now my own ESC/POS UWP driver in c#.  Supports perfectly Serial (actually USB to serial convertors since UWP does not support serial onboard) and IP printer.  No OPOS, no messing with various limitation of namespace around and structured in such a way to get it later onto Xamarin.  Lot of effort in the project, but with great end result.

I would to extend further the interface support.  Next hurdle to take is the support of POS printer with only USB interface under UWP.  But I have no clue to tacle this.  Can you advise me please ?  To be clear : I do not want to use any other driver software from any POS hardware vendor.  Solution should be based on the sending directly ESC/POS commands over the USB wire.

Regards.

**Comment by antalistvan** [comment]
Hi Dirk!

It's sounds wonderful. Do you have any plan to publish it?

We are working on a huge system. Clients are on UWP (plan).

Only one thing stopped us. Our company has around 50+ label printers, POS printers, card terminals, special hardware. These things are only programmable through COM or USB or parallel ports using escape sequences.

The most critical point is how to communicate with these hardware from UWP... 

The "I don't know how it works but maybe, magically doing things appconverter" is not an option.

Plan B: we must stay on winform. Not funny.

**Comment by DirkRoelandt** [comment]
Hi,

Currently we have no plans to publish it.

Guess we are facing similar challenge with hardware interfacing part on UWP.

**Comment by antalistvan** [comment]
Hi,

No problem.

I think this "problem" is not our challenge, it's the UWP stupidity. Nice, new, shiny - and partially - unusable.

Can you share me some "direction" about how can we do something similar to your solution?

**Comment by TerryWarwick** [comment]
Gentlemen,

The challenge with USB connectivity for POSPrinters is that there is currently no standard hardware interface to which the physical printer can bind to a specific driver.  I believe we will have a solution in a future update that will meet your needs in UWP on desktop. 

ESC/POS can also be challenging as manufacturers do not fully adhere to the specification and each device model has its own exceptions.

Terry Warwick
Microsoft

**Comment by DirkRoelandt** [comment]
Hi Terry,

Thanks for the hint, but I want now a solution and without extra dependency.

Having implemented already the flavours of ESC/POS protocol, we really do not want to wait for something a specific driver (most likely just again for a limited numbers of ESC/POS printers neither giving us the control/flexibility we need).   

So do you think it is realistic to further enhanced our current ESC/POS wrapper solution for IP and USB convertors with an extension to WinUSB ?
(i.e. https://msdn.microsoft.com/en-us/library/windows/hardware/ff540174(v=vs.85).aspx, our by extension via c# wrappers like https://github.com/madwizard-thomas/winusbnet)
Like several persons suggested in this thread it would potentially give a better, direct, more economic and appropriate approach for some weird (printer) peripherals mainly on USB (or even plain RS232 - do not want to address this yet in order to keep focus on USB hurdle at hand)

Basically eliminating the perception of limited/no UWP hardware interface support on "desktop" machines.

Awaiting your advise on my query.

Good to you.

Dirk

**Comment by hjgode** [comment]
Hi Dirk

possibly [this](https://msdn.microsoft.com/en-us/library/windows/hardware/dn303343%28v=vs.85%29.aspx) will help you in getting a connection to the printers (or other usb devices with no specific driver).

Josef

**Comment by DirkRoelandt** [comment]
As far my know-how stretches this is not an option.   Eg POS printer are automatically linked via usbprint.sys.  

When trying the suggested approach for the specific device  : 

            string aqs = UsbDevice.GetDeviceSelector(0x0FE6, 0x811E);
            var myDevices = await DeviceInformation.FindAllAsync(aqs);

myDevices yields no results.   Also the CustomUsbDeviceAccess UWP sample doesn't give any result hoped for.

Strange, because the link suggests any USB should be reachable.

**Comment by DenisPitcher** [comment]
I'd like to add to the chorus of people expressing disappointment over the lack of usb support for receipt printers in UWP.  Our clients have a large investment in Epson usb thermal printers as they are far easier to manage than wireless ones which can have config issues.  Seems we'll sadly have to rethink our plans to adopt UWP as a result.

**Comment by TerryWarwick** [comment]
@DenisPitcher,

We started with Network and Bluetooth POSPrinters to support the emergence of mPOS scenarios for tablets.  USB requires special handling due to the lack of hardware interface standardization for POSPrinter.  OPOS got around this by requiring every HW vendor to publish device and model specific service objects.  This is a very painful experience and one that we truly want to bring to par with your typical PC peripherals.  Starting with the APIs that we have available now you will allow you to take advantage of USB support once it is available in Windows desktop, which I can assure you we are doing everything we can to light this up sooner rather than later.

Terry Warwick
Microsoft

**Comment by TerryWarwick** [comment]
This thread has diverged from the purpose of this forum, which is to address issues specific to the samples provided.  Please open a new thread for new issues associated with the POSPrinter sample.  For general issues related to incorporating support for POSPrinter into your application, please post your questions to the "[Developing Universal Windows apps](https://social.msdn.microsoft.com/Forums/windowsapps/en-US/home?forum=wpdevelop&filter=alltypes&sort=lastpostdesc)" forum on MSDN.

This thread is now locked.
